File: src/router/repository/lookup.ts
import { type Origin } from '@prisma/client'
import type { NextFunction, Request, Response } from '@tinyhttp/app'
import { prisma } from 'database.js'

type LookupResponse = Response & {
	locals: {
		query: string;
	};
}

export function middleware(request: Request, response: LookupResponse, next: NextFunction) {
	const query = request.params.repository
	if (!query) {
		return response.status(400)
			.json({
				message: '400 Bad Request',
				error: 'Missing URL parameter: \':repository\'',
				date: new Date()
			})
	}

	response.locals.query = query.toString()
	next()
}

export async function handler(_request: Request, response: LookupResponse) {
	const { query } = response.locals
	const repo = await prisma.repository.findFirst({
		where: {
			slug: query,
			isPruned: false
		},

		include: {
			origin: true
		}
	})

	if (!repo) {
		return response.status(404)
			.json({
				message: '404 Not Found',
				error: 'Repository not found',
				date: new Date()
			})
	}

	const entries = Object.entries(repo)
		.map(([key, value]) => {
			if (key === 'origin') {
				const filtered = Object.fromEntries(Object.entries(value as Origin)
					.filter(([key]) => key !== 'uuid'))

				return [key, filtered]
			}

			return [key, value]
		})
		.filter(([key]) => key !== 'originId' && key !== 'isPruned')

	return response.status(200)
		.json({
			message: '200 Successful',
			date: new Date(),
			// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
			data: {
				...Object.fromEntries(entries),
				refs: {
					packages: `${$product.api_endpoint}/jailbreak/repository/${repo.slug}/packages`
				}
			}
		})
}

File: api/src/routes/repository/lookup.rs
use crate::prisma::repository;
use crate::utility::json_respond;
use crate::{db::prisma, utility::merge_json};

use serde_json::json;
use tide::{
	Request, Result,
	StatusCode::{BadRequest, NotFound, Ok as OK},
};
use tokio::runtime::Builder;

pub async fn repository_lookup(req: Request<()>) -> Result {
	let query = match req.param("repository") {
		Ok(query) => query.to_string(),
		Err(_) => {
			return Ok(json_respond(
				BadRequest,
				json!({
					"message": "400 Bad Request",
					"error": "Missing URL parameter: \':repository\'",
					"date": chrono::Utc::now().to_rfc3339(),
				}),
			));
		}
	};

	let repository = Builder::new_multi_thread()
		.enable_all()
		.build()
		.unwrap()
		.block_on(async move {
			return prisma()
				.await
				.repository()
				.find_first(vec![
					repository::slug::equals(query.to_string()),
					repository::is_pruned::equals(false),
				])
				.with(repository::origin::fetch())
				.exec()
				.await
				.unwrap();
		});

	match repository {
		Some(repository) => {
			let slug = repository.slug.clone();
			let repository = merge_json(
				repository,
				json!({
					"refs": {
						"packages": format!("{}/jailbreak/repository/{}/packages", env!("CANISTER_API_ENDPOINT"), slug),
					}
				}),
			);

			return Ok(json_respond(
				OK,
				json!({
					"message": "200 Successful",
					"date": chrono::Utc::now().to_rfc3339(),
					"data": repository,
				}),
			));
		}
		None => {
			return Ok(json_respond(
				NotFound,
				json!({
					"message": "404 Not Found",
					"error": "Repository not found",
					"date": chrono::Utc::now().to_rfc3339(),
				}),
			))
		}
	}
}
